api/client/src/main/java/org/projectnessie/client/auth/oauth2/AuthorizationCodeFlow.java

```diff
@@ -0,0 +1,260 @@
+    this.flowTimeout = params.getAuthorizationCodeFlowTimeout();
+    authCodeFuture = requestFuture.thenApply(this::extractAuthorizationCode);
+    tokensFuture = authCodeFuture.thenApply(this::fetchNewTokens);
+    closeFuture.thenRun(this::doClose);
+    server =
+        createServer(params.getAuthorizationCodeFlowWebServerPort().orElse(-1), this::doRequest);
+    state = OAuth2Utils.randomAlphaNumString(STATE_LENGTH);
+    redirectUri =
+        String.format("http://localhost:%d%s", server.getAddress().getPort(), CONTEXT_PATH);
+    URI authEndpoint = params.getAuthEndpoint().orElseThrow(IllegalStateException::new);
+    authorizationUri =
+        new UriBuilder(authEndpoint.resolve("/"))
+            .path(authEndpoint.getPath())
+            .queryParam("response_type", "code")
+            .queryParam("client_id", params.getClientId())
+            .queryParam("scope", params.getScope().orElse(null))
+            .queryParam("redirect_uri", redirectUri)
+            .queryParam("state", state)
+            .build();
+    LOGGER.debug("Authorization Code Flow: started, redirect URI: {}", redirectUri);
+  }
+
+  @Override
+  public void close() {
+    closeFuture.complete(null);
+  }
+
+  private void doClose() {
+    inflightRequestsPhaser.arriveAndAwaitAdvance();
+    LOGGER.debug("Authorization Code Flow: closing");
+    server.stop(0);
+    // don't close the console, it's not ours
+  }
+
+  private void abort() {
+    tokensFuture.cancel(true);
+    authCodeFuture.cancel(true);
+    requestFuture.cancel(true);
+  }
+
+  public Tokens fetchNewTokens() {
+    console.println(MSG_PREFIX + "Browse to the following URL to continue:");
+    console.println(MSG_PREFIX + authorizationUri);
+    console.flush();
+    try {
+      Tokens tokens = tokensFuture.get(flowTimeout.toMillis(), TimeUnit.MILLISECONDS);
+      console.println(
+          MSG_PREFIX + "Authentication successful: you can close the browser page now.");
+      console.flush();
+      return tokens;
+    } catch (TimeoutException e) {
+      console.println(MSG_PREFIX + "Timed out waiting for authorization code.");
+      console.flush();
+      abort();
+      throw new RuntimeException(e);
+    } catch (InterruptedException e) {
+      abort();
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException e) {
+      abort();
+      Throwable cause = e.getCause();
+      console.println(MSG_PREFIX + "Authentication failed: " + cause.getMessage());
+      console.flush();
+      if (cause instanceof HttpClientException) {
+        throw (HttpClientException) cause;
+      }
+      throw new RuntimeException(cause);
+    }
+  }
+
+  private void doRequest(HttpExchange exchange) {
+    LOGGER.debug("Authorization Code Flow: received request");
+    inflightRequestsPhaser.register();
+    requestFuture.complete(exchange);
+    tokensFuture
+        .handle((tokens, error) -> doResponse(exchange, error))
```

Comment:
Heads up: I added today a test for bad codes and realized I was wrongly using `whenComplete` here, which propagates the error from `tokensFuture` and thus prevents `.thenRun(inflightRequestsPhaser::arriveAndDeregister)` from executing. The result is that the flow gets stuck in `close()`. Using `handle` instead fixes that.

Reply:
Nice catch! However, by the same logic, we should probably use `.whenComplete()` for the `close` and `arriveAndDeregister` calls below... WDYT?

Reply:
I think we are fine: `doResponse` does not throw so `handle` doesn't either.

Reply:
OK let's change it to err on the safe side.